File: src/Usuario.hs
module Usuario (Usuario (..), TipoUsuario (..), cadastrarUsuario, login, modificarUsuario) where

data TipoUsuario = ProductOwner | ScrumMaster | DevTeam
  deriving (Show, Eq)

data Usuario = Usuario
  { usuarioId :: Int,
    usuarioNome :: String,
    usuarioEmail :: String,
    usuarioSenha :: String,
    usuarioPapel :: TipoUsuario,
    usuarioEmpresaId :: Int
  }
  deriving (Show, Eq)

cadastrarUsuario :: IO Usuario
cadastrarUsuario = do
  putStrLn "Digite o ID do Usuário:"
  id <- readLn
  putStrLn "Digite o Nome do Usuário:"
  nome <- getLine
  putStrLn "Digite o Email do Usuário:"
  email <- getLine
  putStrLn "Digite a senha do Usuário:"
  senha <- getLine
  putStrLn "Selecione o tipo de usuário (1 para Product Owner, 2 para Scrum Master, 3 para Dev Team):"
  tipo <- getLine
  let papel = case tipo of
        "1" -> ProductOwner
        "2" -> ScrumMaster
        "3" -> DevTeam
        _ -> DevTeam -- Padrão para DevTeam se entrada inválida
  putStrLn "Digite o ID da Empresa:"
  empresaId <- readLn
  let usuario = Usuario id nome email senha papel empresaId
  putStrLn $ "Usuário cadastrado: " ++ show usuario
  return usuario

login :: [Usuario] -> IO (Maybe Usuario)
login usuarios = do
  putStrLn "Digite o Email do Usuário:"
  email <- getLine
  let usuario = filter (\u -> usuarioEmail u == email) usuarios
  if null usuario
    then do
      putStrLn "Usuário não encontrado."
      return Nothing
    else do
      let user = head usuario
      putStrLn "Digite a Senha:"
      senha <- getLine
      if senha == usuarioSenha user
        then do
          putStrLn $ "Login bem-sucedido! Bem-vindo, " ++ usuarioNome user
          return $ Just user
        else do
          putStrLn "Senha incorreta."
          return Nothing

modificarUsuario :: Usuario ->  IO Usuario
modificarUsuario usuario = do
  putStrLn "\nDados do Usuário:"
  putStrLn $ "1. Nome: " ++ usuarioNome usuario
  putStrLn $ "2. Email: " ++ usuarioEmail usuario
  putStrLn $ "3. Senha: " ++ usuarioSenha usuario
  putStrLn $ "4. ID da Empresa(Não pode ser modificado): " ++ show (usuarioEmpresaId usuario)
  putStrLn "Escolha o número do campo que deseja modificar (ou 0 para sair):"
  escolha <- getLine
  case escolha of
    "1" -> do
      putStrLn "Digite o novo nome:"
      novoNome <- getLine
      modificarUsuario usuario { usuarioNome = novoNome }  
    "2" -> do
      putStrLn "Digite o novo email:"
      novoEmail <- getLine
      modificarUsuario usuario { usuarioEmail = novoEmail }  
    "3" -> do
      putStrLn "Digite a nova senha:"
      novaSenha <- getLine
      modificarUsuario usuario { usuarioSenha = novaSenha }  
    "0" -> return usuario  
    _   -> do
      putStrLn "Opção inválida, tente novamente."
      modificarUsuario usuario  -- Chama novamente em caso d

File: prolog/src/usuario.pl
:- module(usuario, [usuario/6, usuario_papel/2, cadastrar_usuario/0, login/2, modificar_usuario/2, usuario_empresa_id/2]).

:- dynamic usuario/6.


papel_string(1, 'Product Owner').
papel_string(2, 'Scrum Master').
papel_string(3, 'Dev Team').

% Função para cadastrar um usuário
cadastrar_usuario :- 
    writeln('Digite o ID do Usuário:'),
    read_line_to_string(user_input, IdStr),
    atom_string(Id, IdStr),
    writeln('Digite o Nome do Usuário:'),
    read_line_to_string(user_input, Nome),
    writeln('Digite o Email do Usuário:'),
    read_line_to_string(user_input, Email),
    writeln('Digite a senha do Usuário:'),
    read_line_to_string(user_input, Senha),
    writeln('Selecione o tipo de usuário (1 para Product Owner, 2 para Scrum Master, 3 para Dev Team):'),
    read_line_to_string(user_input, TipoStr),
    atom_number(TipoStr, Papel),  % Papel agora é numérico
    (Papel >= 1, Papel =< 3 -> true; Papel = 3),  % Padrão para Dev Team se entrada inválida
    writeln('Digite o ID da Empresa:'),
    read_line_to_string(user_input, EmpresaIdStr),
    atom_string(EmpresaId, EmpresaIdStr),
    assert(usuario(Id, Nome, Email, Senha, Papel, EmpresaId)),
    writeln('Usuário cadastrado: '),
    writeln(usuario(Id, Nome, Email)).

% Função para fazer login de um usuário
login(Usuario, PapelNum) :-
  writeln('Digite o Email do Usuário:'),
  read_line_to_string(user_input, Email),
  findall(usuario(Id, Nome, Email, Senha, PapelNum, EmpresaId), usuario(Id, Nome, Email, Senha, PapelNum, EmpresaId), Usuarios),
  (   Usuarios = [] ->
    writeln('Usuário não encontrado.'),
    Usuario = none,
    PapelNum = 0
  ;   Usuarios = [usuario(Id, Nome, Email, Senha, PapelNum, EmpresaId)] ->
    writeln('Digite a Senha:'),
    read_line_to_string(user_input, SenhaInput),
    (   SenhaInput == Senha ->
      format('Bem-vindo, ~w!~n', [Nome]),
      Usuario = usuario(Id, Nome, Email, Senha, PapelNum, EmpresaId)
    ;   writeln('Senha incorreta.'),
      Usuario = none,
      PapelNum = 0
    )
  ).

% Função para modificar os dados de um usuário
modificar_usuario(usuario(Id, Nome, Email, Senha, Papel, EmpresaId), UsuarioModificado) :-
    writeln('\nDados do Usuário:'),
    format('1. Nome: ~w~n', [Nome]),
    format('2. Email: ~w~n', [Email]),
    format('3. Senha: ~w~n', [Senha]),
    format('4. ID da Empresa (Não pode ser modificado): ~w~n', [EmpresaId]),
    writeln('Escolha o número do campo que deseja modificar (ou 0 para sair):'),
    read_line_to_string(user_input, EscolhaStr),
    atom_number(EscolhaStr, Escolha),
    (Escolha = 1 ->
        writeln('Digite o novo nome:'),
        read_line_to_string(user_input, NovoNome),
        UsuarioModificado = usuario(Id, NovoNome, Email, Senha, Papel, EmpresaId);
     Escolha = 2 ->
        writeln('Digite o novo email:'),
        read_line_to_string(user_input, NovoEmail),
        UsuarioModificado = usuario(Id, Nome, NovoEmail, Senha, Papel, EmpresaId);
     Escolha = 3 ->
        writeln('Digite a nova senha:'),
        read_line_to_string(user_input, NovaSenha),
        UsuarioModificado = usuario(Id, Nome, Email, NovaSenha, Papel, EmpresaId);
     Escolha = 0 ->
        UsuarioModificado = usuario(Id, Nome, Email, Senha, Papel, EmpresaId);
     writeln('Opção inválida, tente novamente.'),
     modificar_usuario(usuario(Id, Nome, Email, Senha, Papel, EmpresaId), UsuarioModificado)
    ).

% Função para obter o ID da empresa de um usuário
usuario_empresa_id(Usuario, EmpresaId) :- 
    usuario(Usuario, _, _, _, _, EmpresaId).

% Função para verificar o papel de um usuário
usuario_papel(Usuario, Papel) :- 
    usuario(Usuario, _, _, _, PapelNum, _),  
    format('Usuário: ~w, Papel numérico: ~w~n', [Usuario, PapelNum]),
    Papel = PapelNum.